h2o-algos/src/main/java/hex/glm/GLMModel.java

```diff
@@ -1263,18 +1279,40 @@ public int nclasses() {
       return res;
     }
 
-    public Submodel pickBestModel() {
-      int bestId = 0;
-      Submodel best = _submodels[0];
-      for(int i = 1; i < _submodels.length; ++i) {
-        Submodel sm = _submodels[i];
-        if(!(sm.devianceTest > best.devianceTest) && sm.devianceTrain < best.devianceTrain){
-          bestId = i;
-          best = sm;
+    public double selectBestSubmodel(double trainDev, double testDev, double[] xval_test_deviances, int alphaIndex,
+                                     int lambdaIndex, int submodelIdx, double bestDevValue, boolean setAlphaLambdaIdx) {
+      // if testDev exists, use it above everything else;
+      // else if xval exists, use it above trainDev
+      // use trainDev if that is all that is available
+      boolean foundBetterDev = false;
+      double bestValue = bestDevValue;
+      if (!Double.isNaN(testDev) && (testDev < bestDevValue)) { // use validation metrics
```

Comment:
doesn't look right: if we have a `testDev` but it's larger that `bestDevValue`, then we fall back to xval or `trainDev`, which itself may be lower that `bestDevValue`: so we provide sth with high precedence, but the lower precedence deviance will be used at the end...
Looks like the logic should be split into 2 steps:
1. select the deviance that takes precedence: `testDev`, `trainDev`, `xvalDev`.
2. compare its value with `bestDev`.

Reply:
@sebhrusen That is bug.  I need to fix it.

Reply:
I removed the whole bestValue thing and just use what was there before.